### PR TITLE
[WIP] Reset dataset programatically

### LINKED ISSUE #34: Add the ability to reset a dataset programatically
Hi!  I heard the podcast on this and liked the ideas behind it, and at the same time had a component that I needed to add infinite scroll to, so I thought I'd give it a whirl.  I'm on my second day now and I still haven't got this basically working in my app.  Hate create an issue for it but was unsure of what other venue I had. If I stick with this library I'd be happy to PR any shortcomings. 

So here it goes...

1. No clear way to reset the impaginator dataset.  
Since I'm a search component, my dataset changes as the user changes the search query.  Whenever the user changes the search.  I couldn't find a way to reset things at first.  The dataset seems to only be recreated when the pagination parameters change, which I didn't want to do.  I finally settled on just creating the dataset myself and passing it as a component attribute, which meant I also had to pass the dataset state as an attribute (since the dataset has an observer that sets it).  Anyways, that seems to get me farther, but also seems hackish
2. My list is rendering before the promise resolves from fetch
my fetch function is called, in which I make a query call to the ember store and return the resulting promise.  I add a `then` to that promise which sets the `stats.totalPages` and sets the result data on my own component (so that I can fill in decorating information on the UI about results), and I return an array of results
Here's the order I see: 

* `fetch` is called (returning the ember-data query promise)
* list is rendered (with a bunch of undefineds)
* `then` function runs and returns results
* nothing re-rendered

Any guidance around this would be appreciated.  I can post some code as well, but I wanted to make sure my scenario is one that makes sense.

### PR DESCRIPTION
**Pull request by flexyford**
This introduces an API feature which allows the user to reset, refresh or setReadOffset on the dataset both from the handlebars template, as well as from a controller / route. 
Fixes #34

> Note: This PR only provides an API to set the readOffset on the dataset explicitly. In order to reset all fetched records in the dataset, I refer you to https://github.com/thefrontside/ember-impagination/issues/34#issuecomment-169219312. 
> TLDR: If you want to fully reset the `impagination-dataset` to an empty state, you have to pass it a new `fetch` function.

Reasons for resetting or refreshing a dataset:
1. Since the dataset only resets when `{{impagination-dataset}}` parameters change, there is no clear way to reset the dataset. When we want to fetch data from the backend with a query which does not rerender the route, we must manually fire a dataset reset/refresh.
2. To comply with a planned feature to filter-records on the frontend, we will want to manually refresh the dataset to apply filters on records which may have changed. Since filtering data on the frontend does not rerender the route, we must manually fire a reset/refresh.

New API features:

``` javascript
actions: {
  reset: function()    // sets the dataset to readOffset `0`
  refresh: function() // sets the dataset to the currentReadOffset
  setReadOffset: function(offset) // sets the dataset to passed-in `offset`
}
```
### Refreshing Dataset Examples

In this example we refresh the dataset upon updated search queries through the `{{search-pane}}` component.
#### from child components

``` hbs
{{#impagination-dataset fetch=myFetch as |dataset|}}
  {{#search-pane search=(action "performSearch") on-search-results=(action "reset" target=dataset)}}
      {{#ember-collection items=dataset as |record|}}
        {{chat-search-result result=record}}
      {{/ember-collection}}
  {{/search-pane}}
{{/impagination-dataset}}
```
#### from parent route

We use the `on-observe` action, which fires whenever there exists a new dataset to observe. We can then perform actions on the new dataset, which are PASSED-UP TO THE ROUTE, along with the dataset.

``` hbs
{{#impagination-dataset fetch=myFetch on-observe=(action "observeDataset") as |dataset|}}
  {{#search-pane search=(action "performSearch")}}
      {{#ember-collection items=dataset as |record|}}
        {{chat-search-result result=record}}
      {{/ember-collection}}
  {{/search-pane}}
{{/impagination-dataset}}
```

``` javascript
    observeDataset: function(dataset, actions) { // dataset and actions are passed-up
        this.set('dataset', dataset);
        this.set('actions', actions);
    },
   _refreshDataset() {
        let refresh = this.get('actions.refresh');
        refresh.call(this.get('dataset'))
    }
```

TODO / WIP:
- [x] refresh should fire the observer
- [x] setReadOffset must fire observer when we pass-in the current `readOffset` of the dataset
- [ ] Write tests for firing actions from child components

### CODE REVIEW

File: addon/components/impagination-dataset.js [hunk 1]
```diff
@@ -46,6 +47,18 @@ export default Ember.Component.extend({
 
     this.setInitialState();
     this.get('dataset').setReadOffset(this.get('read-offset') || 0);
+  },
+
+  actions: {
+    reset: () => {
```
Comment (Robdel12) [inline addon/components/impagination-dataset.js]: Any reason for this over:
```js
reset() {
  //...
}
```

Comment (flexyford) [inline addon/components/impagination-dataset.js]: Woops, this was not supposed to make it in there. Yet another reason for testing all new feature APIs.

**Comment by flexyford** [comment]
Confused as to why this build is failing. Nothing has really changed between 28a1889
and	 79dc17d

**Comment by flexyford** [comment]
Closing in Favor of #130
